### PR TITLE
feat!: Support custom embedding models in the DocumentIndexClient

### PR DESCRIPTION
**Pull request by Michael-JB**
# Description
- Allow custom embedding models when creating new indexes
- Support new InstructableEmbed embedding strategy
- Update test fixture to cover new configuration
- Update CHANGELOG.md with user-facing changes

Note: this is a breaking change. The Document Index has deprecated `embedding_type` in its index configuration, and it will be removed in due course. This PR proactively migrates away from this deprecated field.

## Before Merging
 - [ ] Review the code changes
    - Unused print / comments / TODOs
    - Missing docstrings for functions that should have them
    - Consistent variable names
    - ...
 - [x] Update `changelog.md` if necessary
 - [ ] Commit messages should contain a semantic [label](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) and the ticket number
   - Consider squashing if this is not the case

### CODE REVIEW

File: CHANGELOG.md
Comment (SebastianNiehusAA) [inline CHANGELOG.md:18]: I think we should add a line like "For examples of the new usage, see the document_index.ipynb" with a link to the notebook to make the migration easier

**Review by SebastianNiehusAA — CHANGES_REQUESTED** [review]

File: tests/connectors/document_index/test_document_index.py
Comment (Michael-JB) [reply tests/connectors/document_index/test_document_index.py:104]: I would counter-argue here - if this test fails as it is now, it wouldn't be a flake, it would be a bug! Using randomization here is a nice way to increase the surface area of this test. Wherever we use this function, we effectively define the representation to be an invariant of the test. Therefore, if the test were to fail, this constraint would no longer hold and we would have found a bug.

We already make use of randomization in our other Document Index client tests, so I would be hesitant to remove it here.

File: CHANGELOG.md
Comment (Michael-JB) [reply CHANGELOG.md:18]: Thanks for the suggestion, will update this :)

File: tests/connectors/document_index/test_document_index.py
Comment (SebastianNiehusAA) [reply tests/connectors/document_index/test_document_index.py:104]: I agree that we should should cover all of the possible representations. My worry is just that by letting it pick randomly, we do not always get all of the representations (therefore potentially passing because the broken representation was not tested).  By giving it to the test parameterized,  we could always make sure that all representations are tested and also get a clear failure message which one broke.

Comment (Michael-JB) [reply tests/connectors/document_index/test_document_index.py:105]: Hmm. After thinking about this a bit more and discussing it with the team, it probably makes sense to introduce separate tests as you suggest. Although I think it would have been fine (given there are only two variants), this way we can be more explicit.

Unfortunately it's not a simple as just parametrizing, as the existing random_index fixture also runs cleanup logic. I've implemented what I think is the best option here using a contextmanager. Hopefully this aligns with what you had in mind!

Comment (SebastianNiehusAA) [reply tests/connectors/document_index/test_document_index.py:105]: Yes, the solution looks in line with what I had in mind 👍

**Review by SebastianNiehusAA — APPROVED** [review]
Lookin good with the new changes. 
Will merge once we get a second opinion from the team, considering that the MR has a breaking change.

**Pull request merged**
